java/arcs/core/data/expression/PaxelParser.kt

```diff
@@ -0,0 +1,292 @@
+        Expression.QueryParameterExpression<Any>(it)
+    }
+
+    private val nestedExpression =
+        -regex("(\\($WS*)") + parser(::paxelExpression) + -regex("($WS*\\))")
+
+    private val unaryOperation =
+        ((token("not ") / token("-")) + ows + parser(::primaryExpression)).map { (token, expr) ->
+            Expression.UnaryExpression(
+                Expression.UnaryOp.fromToken(token.trim()) as Expression.UnaryOp<Any, Any>, expr
+            )
+        }
+
+    private val primaryExpression: Parser<Expression<Any>> =
+        nestedExpression /
+        discreteValue /
+        numberValue /
+        unaryOperation /
+        booleanValue /
+        textValue /
+        functionCall /
+        scopeLookup /
+        query
+
+    private val multiplyExpression = primaryExpression sepBy binaryOp("*", "/")
+    private val additiveExpression = multiplyExpression sepBy binaryOp("+", "-")
+    private val comparativeExpression = additiveExpression sepBy binaryOp("<=", "<", ">=", ">")
+    private val equalityExpression = comparativeExpression sepBy binaryOp("==", "!=")
+    private val andExpression = equalityExpression sepBy binaryOp("and")
+    private val orExpression = andExpression sepBy binaryOp("or")
+
+    private val refinementExpression = orExpression
+
+    private val sourceExpression = scopeLookup / nestedExpression
+
+    private val fromIter = -token("from") + (ows + ident + ows)
+    private val fromIn = -token("in") + (ows + sourceExpression)
+
+    @Suppress("UNCHECKED_CAST")
+    private val fromExpression: Parser<QualifiedExpression> =
+        (fromIter + fromIn).map { (iter, src) ->
+            Expression.FromExpression(null, src as Expression<Sequence<Any>>, iter)
+        }
+
+    private val whereExpression: Parser<QualifiedExpression> =
+        -token("where") + (ows + refinementExpression).map { expr ->
+            Expression.WhereExpression(
+                expr as Expression<Sequence<kotlin.Unit>>, expr as Expression<Boolean>
+            )
+        }
+
+    private val schemaNames = (ident + many(whitespace + ident)).map { (ident, rest) ->
+        setOf(ident) + rest.toSet()
+    }
+
+    private val oCurly = -regex("(\\{$WS*)")
+    private val cCurly = -regex("($WS*\\})")
+    private val comma = -regex("($WS*,$WS*)")
+    private val colon = -regex("($WS*:$WS*)")
+
+    private val newField = (ident + colon + parser(::paxelExpression))
+    private val newFields = optional(newField + many(comma + newField)).map { fields ->
+        fields?.let { (name, expr, otherFields) ->
+            listOf(name to expr) + otherFields
+        } ?: emptyList()
+    }
+    private val newFieldsDecl = oCurly + newFields + cCurly
+
+    private val newExpression =
+        (-token("new") + (whitespace + schemaNames) + ows + newFieldsDecl).map { (names, fields) ->
+            Expression.NewExpression(names, fields)
+        }
+
+    private val selectExprArg = newExpression / refinementExpression
```

Comment:
I think newExpression should be allowed outside of select as well, e.g.

```
particle Foo
  a: reads A {a: Text}
  b: write B {b: Text} = new B {b: a.a}
```

or even: `union(new B {b: "ipsum"}, new B {b: "lorem"})` (though for that we may want a set/list constructor someday...